lib/lightning/kafka_triggers/message_handling.ex

```diff
@@ -0,0 +1,151 @@
+
+  def process_candidate_for(%MessageCandidateSet{} = candidate_set) do
+    Repo.transaction(fn ->
+      candidate_set
+      |> find_candidate_for()
+      |> lock("FOR UPDATE SKIP LOCKED")
+      |> Repo.one()
+      |> case do
+        nil ->
+          nil
+
+        candidate = %{work_order: nil} ->
+          create_work_order(candidate)
+
+        candidate ->
+          maybe_delete_candidate(candidate)
+      end
+    end)
+
+    :ok
+  end
+
+  defp create_work_order(candidate) do
+    %{
+      data: data,
+      metadata: metadata,
+      trigger: %{workflow: workflow} = trigger
+    } = candidate
+
+    data
+    |> Jason.decode()
+    |> case do
+      {:ok, body} ->
+        {:ok, without_run?} = check_skip_run_creation?(workflow.project_id)
+
+        {:ok, %WorkOrder{id: work_order_id}} =
+          WorkOrders.create_for(trigger,
+            workflow: workflow,
+            dataclip: %{
+              body: body,
+              request: metadata,
+              type: :kafka,
+              project_id: workflow.project_id
+            },
+            without_run: without_run?
+          )
+
+        candidate
+        |> TriggerKafkaMessage.changeset(%{work_order_id: work_order_id})
+        |> Repo.update!()
+
+      {:error, _decode_error} ->
+        # TODO Add details from _decode_error?
+
+        processing_data =
+          candidate.processing_data
+          |> Map.merge(%{"errors" => ["Data is not a JSON object"]})
+
+        candidate
+        |> TriggerKafkaMessage.changeset(%{processing_data: processing_data})
+        |> Repo.update!()
+    end
+  end
+
+  defp maybe_delete_candidate(candidate) do
+    if successful?(candidate.work_order), do: Repo.delete(candidate)
+  end
+
+  def successful?(%{state: state}) do
+    state == :success
+  end
+
+  @doc """
+  Find the MessageCandidateSetCandidate for the MessageCandidateSet identified
+  by the MessageCanididateSetID (i.e. trigger_id, topic, key).
+
+  Within the current implementation, this no longer needs to be a public method,
+  but having it as a public method allows for easier testing.
+  """
+  def find_candidate_for(%MessageCandidateSet{
+        trigger_id: trigger_id,
+        topic: topic,
+        key: nil
+      }) do
+    from t in TriggerKafkaMessage,
+      where: t.trigger_id == ^trigger_id and t.topic == ^topic and is_nil(t.key),
+      order_by: t.offset,
+      limit: 1,
+      preload: [:work_order, trigger: [:workflow]]
+  end
+
+  def find_candidate_for(%MessageCandidateSet{
+        trigger_id: trigger_id,
+        topic: topic,
+        key: key
+      }) do
+    from t in TriggerKafkaMessage,
+      where: t.trigger_id == ^trigger_id and t.topic == ^topic and t.key == ^key,
+      order_by: t.offset,
+      limit: 1,
+      preload: [:work_order, trigger: [:workflow]]
+  end
+
+  # Stolen from the `webhooks_controller.ex` file and simplified until
+  # I understand the details (ask Roger).
+  defp check_skip_run_creation?(project_id) do
```

Comment:
@jyeshe As the comment says, I was unclear wrt existing cases if the above is an acceptable implementation?

Reply:
Nice finding Rory. Here the error `:too_many_runs` needs to be differentiated from `:runs_hard_limit`. With the former, Lightning still accepts the creation of workorders and does it without runs (the runs are blocked) while the with latter error no workorder is created.

Reply:
Thanks @jyeshe - I have updated as follows: https://github.com/OpenFn/lightning/pull/2270/files#diff-335ea25efd1cf677f88d6d071a096f13624ce535843b3190a6d537a6e56821d3R145.